knodle/labeler/CheXpert/matching.py

```diff
@@ -0,0 +1,159 @@
+        self.labeler_config = config
+        self.observation2mention_phrases = self.load_phrases(self.labeler_config.mention_data_dir)
+        self.observation2unmention_phrases = self.load_phrases(self.labeler_config.unmention_data_dir)
+
+        # Add CheXpert specific unmention phrases.
+        if chexpert_data:
+            self.add_unmention_phrases()
+
+    def load_phrases(self, phrases_dir: str) -> DefaultDict:
+        """Read in map from observations to phrases for matching."""
+        observation2phrases = defaultdict(list)
+        for phrases_path in os.listdir(phrases_dir):
+            with open(os.path.join(phrases_dir, phrases_path)) as f:
+                for line in f:
+                    phrase = line.strip().replace("_", " ")
+                    observation = phrases_path.replace("_", " ").title()
+                    if line:
+                        observation2phrases[observation].append(phrase)
+
+        return observation2phrases
+
+    def add_unmention_phrases(self) -> None:
+        """Define additional unmentions. This function is custom-made for the CheXpert rules."""
+        cardiomegaly_mentions\
+            = self.observation2mention_phrases[self.labeler_config.cardiomegaly]
+        enlarged_cardiom_mentions\
+            = self.observation2mention_phrases[self.labeler_config.enlarged_cardiomediastinum]
+        positional_phrases = (["over the", "overly the", "in the"],
+                              ["", " superior", " left", " right"])
+        positional_unmentions\
+            = [e1 + e2
+               for e1 in positional_phrases[0]
+               for e2 in positional_phrases[1]]
+        cardiomegaly_unmentions\
+            = [e1 + " " + e2.replace("the ", "")
+               for e1 in positional_unmentions
+               for e2 in cardiomegaly_mentions
+               if e2 not in ["cardiomegaly",
+                             "cardiac enlargement"]]
+        enlarged_cardiomediastinum_unmentions\
+            = [e1 + " " + e2
+               for e1 in positional_unmentions
+               for e2 in enlarged_cardiom_mentions]
+
+        self.observation2unmention_phrases[self.labeler_config.cardiomegaly]\
+            = cardiomegaly_unmentions
+        self.observation2unmention_phrases[self.labeler_config.enlarged_cardiomediastinum]\
+            = enlarged_cardiomediastinum_unmentions
+
+    def overlaps_with_unmention(self,
+                                sentence: bioc.BioCSentence,
+                                observation: str,
+                                start: int,
+                                end: int) -> bool:
+        """Return "True" if a given match overlaps with an unmention phrase."""
+        unmention_overlap = False
+        unmention_list = self.observation2unmention_phrases.get(observation,
+                                                                [])
+        for unmention in unmention_list:
```

Comment:
If this is a time consuming component of the code. Maybe restricting text to start / end first, then searching gives performance gain

Reply:
I am not sure whether I understand what you exactly mean by restricting it to start / end first, but it's not a time consuming component - unmentions do not happen too often

Reply:
Ok, then never mind :D